core/src/main/java/io/confluent/rest/TenantDosFilter.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2025 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.rest;
+
+import static io.confluent.rest.TenantUtils.UNKNOWN_TENANT;
+
+import io.confluent.rest.jetty.DoSFilter;
+import jakarta.servlet.ServletRequest;
+import jakarta.servlet.http.HttpServletRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * DoS filter that implements per-tenant rate limiting.
+ * Extends the base DoSFilter to extract tenant IDs from requests and rate limit by tenant.
+ */
+public class TenantDosFilter extends DoSFilter {
+
+  private static final Logger log = LoggerFactory.getLogger(TenantDosFilter.class);
+  
+  public TenantDosFilter() {
+    super();
+  }
+
+  @Override
+  protected String extractUserId(ServletRequest request) {
+    if (!(request instanceof HttpServletRequest)) {
+      log.warn("Request is not an HttpServletRequest, cannot extract tenant ID");
+      return null;
+    }
+
+    HttpServletRequest httpRequest = (HttpServletRequest) request;
+
+    String tenantId = TenantUtils.extractTenantId(httpRequest);
+    
+    if (tenantId.equals(UNKNOWN_TENANT)) {
+      // If we can't identify the tenant, return null to skip tenant-based rate limiting.
+      // This results in the base DoSFilter to fall back to session-based (if enabled) or
+      // IP-based rate limiting as this is the best we can do in this scenario.
```

Comment:
nit: this comment is helpful in understanding the DoSFilter and how to use it by overloading extractUserId. I like how you mention that returning null is the expected case when we want to fall back to session-based rate tracking, but can we put that info at the top of the function? This is because we also return null on L43. I don't want anyone to miss the description of this part of the behavior, since it's pretty important IMO